crates/protocol/src/types/identifier.rs

```diff
@@ -0,0 +1,111 @@
+//! Minecraft's identifier type
+
+use std::{fmt, str::FromStr};
+
+use bytes::{Buf, BufMut};
+
+use crate::{error, ProtocolType};
+
+const MAX_SIZE: usize = 1 << 15;
+
+/// The way to construct this is by calling [`FromStr`].
+#[derive(Debug, Clone)]
+pub struct Identifier {
+    namespace: String,
+    location: String,
+}
+
+impl Identifier {
+    pub fn namespace(&self) -> &str {
+        &self.namespace
+    }
+
+    pub fn location(&self) -> &str {
+        &self.namespace
+    }
+}
+
+impl fmt::Display for Identifier {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}:{}", self.namespace, self.location)
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum IdentifierError {
+    #[error("Not enough colons")]
+    NotEnoughColons,
+    #[error("Too many colons")]
+    TooManyColons,
+    #[error("The identifier is too long: {0} > {}", MAX_SIZE)]
+    TooLong(usize),
+}
+
+impl FromStr for Identifier {
+    type Err = IdentifierError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if s.len() > MAX_SIZE {
+            Err(IdentifierError::TooLong(s.len()))?;
+        }
+
+        let mut split = s.split(":");
+
+        let left = split.next().unwrap();
+        let right = split.next().ok_or(IdentifierError::NotEnoughColons)?;
```

Comment:
This may still be a valid resource identifier. `minecraft:` prefix is considered optional.

Reply:
Is that still true in modern versions? I always thought that commands just add the prefix on their own

Reply:
Yes, at least wiki.vg [says](https://wiki.vg/Protocol#Identifier) so and it also was useable like this previously in resourcepacks.

Reply:
Done